pkg/network/server.go

```diff
@@ -1087,33 +1184,55 @@ func (s *Server) handleGetAddrCmd(p Peer) error {
 // 2. Send requests for chunk in increasing order.
 // 3. After all requests were sent, request random height.
 func (s *Server) requestBlocks(p Peer) error {
+	if !s.canStartSync.Load() && s.p.Load() == 0 {
+		// then need to request headers first, but P1 is not set yet, wait for more peers
+		return nil
+	}
+	// if restoring node state, then fetch headers up to P1+1 height first
+	needHeaders := !s.canStartSync.Load() && s.chain.HeaderHeight() <= s.p.Load()
+
 	var currHeight = s.chain.BlockHeight()
 	var peerHeight = p.LastBlockIndex()
 	var needHeight uint32
+	var lastRequestedHeight = &s.lastBlocksRequestedHeight
+	if needHeaders {
+		currHeight = s.chain.HeaderHeight()
+		lastRequestedHeight = &s.lastHeadersRequestedHeight
+	}
 	// lastRequestedHeight can only be increased.
 	for {
-		old := s.lastRequestedHeight.Load()
+		old := lastRequestedHeight.Load()
 		if old <= currHeight {
 			needHeight = currHeight + 1
-			if !s.lastRequestedHeight.CAS(old, needHeight) {
+			if !needHeaders && !s.canStartSync.Load() && currHeight < s.p.Load()-s.chain.GetConfig().MaxTraceableBlocks {
+				needHeight = s.p.Load() - s.chain.GetConfig().MaxTraceableBlocks
+			}
+			if !lastRequestedHeight.CAS(old, needHeight) {
 				continue
 			}
 		} else if old < currHeight+(blockCacheSize-payload.MaxHashesCount) {
 			needHeight = currHeight + 1
 			if peerHeight > old+payload.MaxHashesCount {
 				needHeight = old + payload.MaxHashesCount
-				if !s.lastRequestedHeight.CAS(old, needHeight) {
+				if !lastRequestedHeight.CAS(old, needHeight) {
 					continue
 				}
 			}
 		} else {
 			index := mrand.Intn(blockCacheSize / payload.MaxHashesCount)
 			needHeight = currHeight + 1 + uint32(index*payload.MaxHashesCount)
+			if !needHeaders && !s.canStartSync.Load() && currHeight < s.p.Load()-s.chain.GetConfig().MaxTraceableBlocks {
+				needHeight = s.p.Load() - s.chain.GetConfig().MaxTraceableBlocks + 1 + uint32(index*payload.MaxHashesCount)
+			}
 		}
 		break
 	}
 	payload := payload.NewGetBlockByIndex(needHeight, -1)
-	return p.EnqueueP2PMessage(NewMessage(CMDGetBlockByIndex, payload))
+	if needHeaders {
```

Comment:
Why don't we make a `requestHeaders` method?

Reply:
A thought about it, and there are two reasons:
1. There's a lot of code duplication between these two methods.
2. We don't need to request headers during the standard sync process, only when state sync is going. During state sync we need to fetch headers first, and only after that blocks can be fetched. So if we still don't reach `p1` headers height then we need to request headers instead of blocks (so everywhere where `requestBlocks` is called, we'll need to call `requestHeaders` instead).

Reply:
After all, it's really better to create a separate method for it, because we actually don't have cache for headers. So fixed.